text/0000-dyn-upcasting.md

```diff
@@ -0,0 +1,248 @@
+
+Consider a flat layout algorithm for a type `T` and a trait `Tr` as follows:
+
+1. Create a tree of all the supertraits of this `TraitRef`, duplicate for the cyclic cases.
+2. Traverse the tree in post-order, for each `TraitRef`, 
+   1. if it has no supertrait, emit a header part, including `MetadataDropInPlace`, `MetadataSize`, `MetadataAlign` items.
+   2. emit all its associated functions as either `Method` or `Vacant` entries.
+
+Given `trait A(n+1): Bn + Cn {}, trait Bn: An { fn bn(&self); }, trait Cn: An { fn cn(&self); }`, the vtable for An will contain 2^n DSAs.
+
+## Why not adopt a "pointer-based" vtable layout?
+
+In this stable, every trait would have their own vtable, and we would embed links to each of them in the subtrait's vtable. This was rejected as it would be less efficient for the single-inheritance case, which is common.
+
+# Prior art
+[prior-art]: #prior-art
+
+Other languages permit upcasting in similar scenarios.
+
+C++ permits upcasting from a reference to a class to any of its superclasses. As in Rust, this may require adjusting the pointer to account for multiple inheritance.
+
+Java programs can upcast from an object to any superclass. Since Java is limited to single inheritance, this does not require adjusting the pointer, but this implies that interfaces are harder.
+
+[Haskell `forall` types permit upcasting](https://wiki.haskell.org/Existential_type#Dynamic_dispatch_mechanism_of_OOP).
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- None.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Arbitrary combinations of traits
+
+It would be very useful to support `dyn Trait1 + Trait2` for arbitrary sets of traits. Doing so would require us to decide how to describe the vtable for the combination of two traits. There is an intefaction between this feature and upcasting, because if we support upcasting, then we must be able to handle upcasting from some subtrait to some arbitrary combination of supertraits. For example a `&mut dyn Subtrait`...
+
+```rust=
+trait Subtrait: Supertrait1 + Supertrait2 + Supertrait3
+```
+
+...could be upcast to any of the following:
+
+* `&mut dyn Supertrait1` (covered by this RFC)
+* `&mut dyn Supertrait2` (covered by this RFC)
+* `&mut dyn Supertrait3` (covered by this RFC)
+* `&mut dyn (Supertrait1 + Supertrait2)` (not covered by this RFC)
+* `&mut dyn (Supertrait2 + Supertrait3)` (not covered by this RFC)
+* `&mut dyn (Supertrait1 + Supertrait3)` (not covered by this RFC)
+* `&mut dyn (Supertrait1 + Supertrait2 + Supertrait3)` (not covered by this RFC)
+
+In particular, this implies that we must be able to go from the vtable for `Subtrait` to any of the above vtables. 
+
+Two ways have been proposed thus far to implement a "multi-trait" dyn like `dyn Trait1 + Trait2`...
+
+* as a single, combined vtable
+* as a "very wide" pointer with one vtable per trait
+
```

Comment:
A third option: The vtable for `Trait1 + Trait2` contains entries for the vtables for `Trait1` and `Trait2`. It's an extra layer of indirection, but much smaller and doesn't explode as quickly, while keeping upcasts trivial.

Reply:
That is what I meant by the first option, I think. You wind up needing each subset of the various traits involved -- well, maybe not EVERY subset, since the vtable for `A + B` might have `A` as a prefix. 

I think this is what we will wind up doing, regardless.